files for the js assignment1

abansal26 opened the pull request:
please check the files

abansal26 committed:
files for the js assignment1

rockon1985 commented on calculateFrequency.js:
```diff
@@ -0,0 +1,14 @@
+// Problem 2: Complete the calculateFrequency function that takes lowercase string as input and returns frequency of all english alphabet. (using only array, no in-built function)
+function calculateFrequency(string) {
+  // Write your code here
```
this comment was to guide you. You can remove it once you write your code.

rockon1985 commented on calculateFrequency.js:
```diff
@@ -0,0 +1,14 @@
+// Problem 2: Complete the calculateFrequency function that takes lowercase string as input and returns frequency of all english alphabet. (using only array, no in-built function)
+function calculateFrequency(string) {
+  // Write your code here
+  var l, i;
```
what is l , i ?? name the variable correctly according to what they represent!

rockon1985 commented on flatten.js:
```diff
@@ -0,0 +1,15 @@
+function flatten(unflatObject) {
+  // Write your code here
+  var outp = {};
+  var x = unflatObject;
```
variables are very badly named! Always name the variables according to what they store. eg: `key, unflatObjectClone`, etc

rockon1985 reviewed (changes requested):
- Name the variables precisely
- Indentation should be 2 spaces per block
- Remove unwanted comments

abansal26 committed:
files for the js assignment1 with changes

abansal26 commented:
Changes added

sakodiya commented on calculateFrequency.js:
```diff
@@ -1,14 +1,17 @@
-  var outp = {};
-  for (i = 0; i < l; i++) {
+  var length;
+
```
@abansal26 Remove the extra line.

sakodiya commented on calculateFrequency.js:
```diff
@@ -1,14 +1,17 @@
+  length = string.length;
+  var output = {};
+
+  for (var i = 0; i < length; i++) {
```
@abansal26 change the var name 'i'

sakodiya commented on calculateFrequency.js:
```diff
@@ -1,14 +1,17 @@
+  for (var i = 0; i < length; i++) {
     var r = string.charAt(i);
+
     if (r >= "a" && r <= "z") {
```
@abansal26 change the var name 'r'

sakodiya commented on secondlargest.js:
```diff
@@ -1,17 +1,18 @@
-  // Write your code here
-  var p1, p2, l;
-  l = array.length;
+  var p1, p2, length;
```
@abansal26 please change the variable names.

abansal26 committed:
files for the js assignment1 with new changes

abansal26 commented:
@sakodiya @ketanTechracers  
Please review the new changes

rockon1985 commented on calculateFrequency.js:
```diff
@@ -0,0 +1,17 @@
+function calculateFrequency(string) {
+  var length;
+
+  length = string.length;
```
If you are assigning the value in the very next line, its better to initialise and assign in a single line only.

rockon1985 commented on secondlargest.js:
```diff
@@ -0,0 +1,18 @@
+  var largest, secondlargest, length;
+
+  length = array.length;
+  largest = secondlargest = 0;
```
declare and initialisation can be done in the same statement.

rockon1985 commented on secondlargest.js:
```diff
@@ -0,0 +1,18 @@
+// Problem 1: Complete the secondLargest function which takes in an array of numbers in input and return the second biggest number in the array. (without using sort)?
+function secondLargest(array) {
+  var largest, secondlargest, length;
```
variable name are supposed to be camelCased. name it `secondLargest`

rockon1985 commented on flatten.js:
```diff
@@ -0,0 +1,16 @@
+
+  for (var iteration in unflatObject) {
+    if (typeof unflatObject[iteration] == "object") {
+      var fo = flatten(unflatObject[iteration]);
```
fo ? name it correctly.

abansal26 committed:
files for the js assignment1 with new changes

abansal26 commented:
@ketanTechracers  @sakodiya  
Please review

sakodiya commented on secondlargest.js:
```diff
@@ -1,18 +1,17 @@
 function secondLargest(array) {
-  var largest, secondlargest, length;
+  var largest =0, secondLargest = 0, length;
 
```
@abansal26 Please remove the extra line in all programs.

sakodiya commented on secondlargest.js:
```diff
@@ -1,18 +1,17 @@
+  var largest =0, secondLargest = 0, length;
 
   length = array.length;
-  largest = secondlargest = 0;
```
@abansal26 **Second largest program** is wrong because it gives incorrect response for **negative numbers**.
In case  [-23, -1, -3, 56] as it should give -1 as output. But your code return 0.

sakodiya commented on calculateFrequency.js:
```diff
@@ -1,8 +1,6 @@
+  var length = string.length;
   var output = {};
 
   for (var iteration = 0; iteration < length; iteration++) {
```
@abansal26 replace iteration with index.

harif-gr replied:
@abansal26 name the variables according to what they do or are meant for.
For example here the for loop does the iteration and not the variable, the variable instead defines the current index for the running iteration. Keep this in mind for all the variables in this PR and in future.

sakodiya commented on calculateFrequency.js:
```diff
@@ -1,8 +1,6 @@
 // Problem 2: Complete the calculateFrequency function that takes lowercase string as input and returns frequency of all english alphabet. (using only array, no in-built function)
 function calculateFrequency(string) {
```
@abansal26 Please look into the program logic. It gives an incorrect response when the input contains capital letters.

harif-gr replied:
@abansal26 why is this comment not acted upon? And if you have already resolved the issue reply to the comment saying done and preferably the commit in which you corrected it.
All comments addressed to you should be replied to with the appropriate status so that the conversation is complete.

abansal26 committed:
changes done as suggested by shubham

harif-gr commented on secondlargest.js:
```diff
@@ -0,0 +1,17 @@
+// Problem 1: Complete the secondLargest function which takes in an array of numbers in input and return the second biggest number in the array. (without using sort)?
+function secondLargest(array) {
+  var largest =0, secondLargest = 0, length;
```
1. Be consistent in your formatting **ALWAYS**! The equals sign has a space after it in one instance and not in one.
2. Follow the guidelines from the link provided on your trello card for whether there should be a space or not, and for all other such instances.

abansal26 commented:
@sakodiya @harif-techracers I completed the changes suggested i.e. changed iteration to index, Tidied up the program and added the logic that will make the programs work in suggested cases. Please review the changes

harif-gr commented on flatten.js:
```diff
@@ -1,16 +1,13 @@
-    if (typeof unflatObject[iteration] == "object") {
-      var flattenObject = flatten(unflatObject[iteration]);
-
+  for (var index in unflatObject) {
```
@abansal26 naming the variable "index" was an example suitable for when for loop iterates on something which has indices for example an array. This is a for each loop being used to iterate over an object, I am sure index is not what it should be named.

abansal26 replied:
Changed index to iteration. Commit ID - 49d27bbc7e17b74bfa261b8f0b29e2d0c364fe83

harif-gr commented on secondlargest.js:
```diff
@@ -1,17 +1,21 @@
-
   return secondLargest;
 }
+
```
@abansal26 Remove extra line

abansal26 replied:
Already removed. Please check the full program.

harif-gr commented on flatten.js:
```diff
@@ -1,16 +1,13 @@
+        output[index + "." + count] = flattenObject[count];
       }
-    } else output[iteration] = unflatObject[iteration];
+    } else output[index] = unflatObject[index];
```
It's good to encapsulate even a single line if, else, for in braces as it improved readability and if not that it definitely should always at least be on the next line with proper indentation.

abansal26 replied:
Changed as suggested. Commit ID - 49d27bbc7e17b74bfa261b8f0b29e2d0c364fe83

harif-gr commented on flatten.js:
```diff
@@ -1,16 +1,13 @@
-      var flattenObject = flatten(unflatObject[iteration]);
-
+  for (var index in unflatObject) {
+    if (typeof unflatObject[index] == "object") {
```
Could you explain what happens here if unflatObject[index] contains an array?

abansal26 replied:
It works. The codepen link takes a array as input as an array is also an object.
Codepen link :- https://codepen.io/abansal26/pen/VzdzvJ/

harif-gr reviewed (changes requested):
Reply to comments individually and make changes where required.

abansal26 committed:
changed index to iteration

abansal26 committed:
moved the files to folder assignment 1

harif-gr commented on assignment 1/flatten.js:
```diff
@@ -0,0 +1,14 @@
+// Problem 3: Complete the flatten function that takes a JS Object, returns a JS Object in flatten format (compressed)
+function flatten(unflatObject) {
+  var output = {};
+  for (var iteration in unflatObject) {
```
@abansal26 this is hilarious! It seems there are only 2 names for a variable used in iteration. Either index or iteration.
We suggested to name variables properly with @sakodiya giving an example to change to index, you changed all to index, to your luck it was relevant in all places but this. I comment you to change it and you revert back to iteration. :D
Don't be a doer be an engineer, read carefully, understand what the comment means, think whether your fix actually resolves that comment?

abansal26 replied:
I now understand what he meant. In this case as we are traversing through object i am now using property as variable to traverse through it. Commit ID - db0d8f6dc6080ddc97bc8d2dc7547bef4105a76f

abansal26 committed:
Changed iteration to property as it is most relevent to traverse through object
